src/olympia/editors/models.py

```diff
@@ -668,6 +670,129 @@ def all_users_by_score(cls):
+    version = models.OneToOneField(
+        Version, on_delete=models.CASCADE, primary_key=True)
+    uses_custom_csp = models.BooleanField(default=False)
+    uses_native_messaging = models.BooleanField(default=False)
+    uses_content_script_for_all_urls = models.BooleanField(default=False)
+    average_daily_users = models.PositiveIntegerField(default=0)
+    approved_updates = models.PositiveIntegerField(default=0)
+    verdict = models.PositiveSmallIntegerField(
+        choices=amo.AUTO_APPROVAL_VERDICT_CHOICES,
+        default=amo.NOT_AUTO_APPROVED)
+
+    def calculate_verdict(
+            self, max_average_daily_users=0, min_approved_updates=0,
+            dry_run=False):
+        """Calculate the verdict for this instance based on the values set
+        on it and the current configuration.
+
+        Return a dict containing more information about what critera passed
+        or not."""
+        if dry_run:
+            success_verdict = amo.WOULD_HAVE_BEEN_AUTO_APPROVED
+            failure_verdict = amo.WOULD_NOT_HAVE_BEEN_AUTO_APPROVED
+        else:
+            success_verdict = amo.AUTO_APPROVED
+            failure_verdict = amo.NOT_AUTO_APPROVED
+
+        # We need everything in that dict to be False for verdict to be
+        # successful.
+        verdict_info = {
+            'uses_custom_csp': self.uses_custom_csp,
+            'uses_native_messaging': self.uses_native_messaging,
+            'uses_content_script_for_all_urls':
+                self.uses_content_script_for_all_urls,
+            'too_many_average_daily_users':
+                self.average_daily_users >= max_average_daily_users,
+            'too_few_approved_updates':
+                self.approved_updates < min_approved_updates,
+        }
+        if any(verdict_info.values()):
+            self.verdict = failure_verdict
+        else:
+            self.verdict = success_verdict
+
+        return verdict_info
+
+    @classmethod
+    def check_uses_custom_csp(cls, version):
+        def _check_uses_custom_csp_in_file(file_):
+            try:
+                validation = file_.validation
+            except FileValidation.DoesNotExist:
+                raise AutoApprovalNoValidationResultError()
+            validation_data = json.loads(validation.validation)
+            return any('MANIFEST_CSP' in message['id']
+                       for message in validation_data.get('messages', []))
+        return any(_check_uses_custom_csp_in_file(file_)
+                   for file_ in version.all_files)
+
+    @classmethod
+    def check_uses_native_messaging(cls, version):
+        return any('nativeMessaging' in file_.webext_permissions_list
+                   for file_ in version.all_files)
+
+    @classmethod
+    def check_uses_content_script_for_all_urls(cls, version):
+        return any(p.name == 'all_urls' for file_ in version.all_files
+                   for p in file_.webext_permissions)
```

Comment:
fyi: I'm mid-refactoring how all_urls is done so they'll likely be a more efficient way of checking if it uses all_urls than calling `.webext_permissions` soon (this should still work though).